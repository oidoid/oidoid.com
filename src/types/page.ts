/** All data associated with a Markdown file. */
export type Page = {
  readonly meta: Meta
  /** **Unsanitized** HTML. */
  readonly html: string
  readonly type: PageType
  /** Relative canonical URL. */
  readonly url: string
}

export type PageType =
  | 'Homepage'
  | 'Index'
  | 'Log'
  | 'Note'
  | 'Profile'

/** Metadata at the top of a markdown file. */
export type Meta = {
  readonly dateModified: string
  readonly datePublished: string
  readonly keywords: string[]
  readonly headline: string
  readonly hideTitle: boolean
  readonly image?: string | undefined
  readonly title: string
}
