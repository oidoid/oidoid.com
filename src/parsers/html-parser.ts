import {
  marked,
  RendererExtension,
  RendererExtensionFunction,
  RendererThis,
  Token,
  TokenizerExtension,
  TokenizerThis,
  Tokens,
} from 'marked'
import { escape } from 'std/html/mod.ts'
import { Meta } from '../types/page.ts'
import { ElementID } from '../utils/element-util.ts'

export async function parseHTML(meta: Meta, markdown: string): Promise<string> {
  marked.use({ extensions: [new SectionExtension(meta)] }, {
    renderer: {
      code(code) {
        // Unindent fences. Fences are indented so the section parser doesn't
        // need to discriminate headings within fences. Four space-prefix blocks
        // without the ``` openers and closers should work too.
        // deno-lint-ignore no-regex-spaces
        return `<pre>${(code.text.replace(/^  /mg, ''))}</pre>`
      },
    },
  })
  const html = await marked.parse(
    `# ${meta.hideTitle ? '' : meta.title}\n${markdown}`, // Inject the title from meta.
    { gfm: true },
  )
  return `<div class='markdown'>${html}</div>`
}

type HeadingToken = Tokens.Generic & {
  heading: { level: number; text: string | undefined; tokens: Token[] }
}

class SectionExtension implements TokenizerExtension, RendererExtension {
  readonly name = 'section'
  readonly level = 'block'

  constructor(meta: Meta) {
    this.renderer = function (this: RendererThis, token: Tokens.Generic) {
      return renderSection.call(
        this,
        <HeadingToken> token,
        meta,
      )
    }
  }

  tokenizer(this: TokenizerThis, src: string): HeadingToken | undefined {
    const match = src.match(
      /(?<hash>^#{1,6}) (?<heading>[^\n]*)(?<body>[^]*?)(?=\n#|$)/,
    )
    if (!match?.groups?.hash) return

    return {
      type: 'section',
      raw: match[0],
      heading: {
        level: match.groups.hash.length,
        text: match.groups.heading,
        tokens: match.groups.heading == null
          ? []
          : this.lexer.inline(match.groups.heading), // Parse the heading.
      },
      tokens: match.groups.body == null
        ? []
        : this.lexer.blockTokens(match.groups.body), // Parse the body.
    }
  }

  renderer: RendererExtensionFunction = () => undefined
}

function renderSection(
  this: RendererThis,
  token: HeadingToken,
  meta: Meta,
): string {
  let heading
  if (token.heading.level === 1) {
    heading = token.heading.text
      ? `
      <page-title
        modified='${escape(meta.dateModified)}'
        label='${escape(meta.title ?? '')}'
        published='${escape(meta.datePublished)}'
        sublabel='${escape(meta.title ? meta.headline : '')}'></page-title>
    `
      : ''
  } else {
    heading = `
      <section-heading
        id='${ElementID(token.heading.text ?? '')}'
        level=${token.heading.level}
      >
        ${this.parser.parseInline(token.heading.tokens)}
      </section-heading>
    `.trim()
  }
  return `
    <section>
      ${heading}
      <div class="section-body">${this.parser.parse(token.tokens ?? [])}</div>
    </section>
  `
}
