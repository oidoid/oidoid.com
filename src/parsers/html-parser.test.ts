import { assertSnapshot } from 'std/testing/snapshot.ts'
import { Meta } from '../types/page.ts'
import { parseHTML } from './html-parser.ts'

const meta: Meta = {
  dateModified: '2000-01-02',
  datePublished: '2000-01-01',
  headline: 'Headline.',
  hideTitle: false,
  title: 'Title',
  keywords: [],
}

Deno.test('heading with empty body', async (test) => {
  await assertSnapshot(test, parseHTML(meta, '## 2'), { dir: '.' })
})

Deno.test('heading with nonempty body', async (test) => {
  await assertSnapshot(test, parseHTML(meta, '## 2\na'), { dir: '.' })
})

Deno.test('two headings', async (test) => {
  await assertSnapshot(test, parseHTML(meta, '## 2\na\n## 3\nb'), { dir: '.' })
})

Deno.test('h2-6', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      `
## 2
a

### 3
b
c

#### 4
d
e
f

##### 5
g
h
i
j

###### 6
k
l
m
n
o
`,
    ),
    { dir: '.' },
  )
})

Deno.test('fence', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      '\n\
```\n\
a\n\
b\n\
c\n\
```\n',
    ),
    { dir: '.' },
  )
})

Deno.test('indented fence', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      '\n\
```\n\
  a\n\
  b\n\
  c\n\
```\n',
    ),
    { dir: '.' },
  )
})

Deno.test('doubly indented fence', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      '\n\
```\n\
    a\n\
    b\n\
    c\n\
```\n',
    ),
    { dir: '.' },
  )
})

Deno.test('fence with hashes', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      '\n\
```\n\
  ## a\n\
  ## b\n\
  ## c\n\
```\n',
    ),
    { dir: '.' },
  )
})

Deno.test('heading before fence', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      '\n\
## 2\n\
a\n\
\n\
```\n\
b\n\
c\n\
d\n\
```\n',
    ),
    { dir: '.' },
  )
})

Deno.test('heading after fence', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      '\n\
```\n\
a\n\
b\n\
c\n\
```\n\
\n\
## 2\n\
d\n',
    ),
    { dir: '.' },
  )
})

Deno.test('heading, fence, heading', async (test) => {
  await assertSnapshot(
    test,
    parseHTML(
      meta,
      '\n\
## 2\n\
a\n\
\n\
```\n\
b\n\
c\n\
d\n\
```\n\
### 3\n\
e\n',
    ),
    { dir: '.' },
  )
})
