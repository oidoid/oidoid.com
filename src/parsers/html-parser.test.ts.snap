export const snapshot = {};

snapshot[`heading with empty body 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"></div>
    </section>
  
    <section>
      <section-heading
        id='id-2'
        level=2
      >
        2
      </section-heading>
      <div class="section-body"></div>
    </section>
  </div>\`,
}
`;

snapshot[`heading with nonempty body 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"></div>
    </section>
  
    <section>
      <section-heading
        id='id-2'
        level=2
      >
        2
      </section-heading>
      <div class="section-body"><p>a</p>
</div>
    </section>
  </div>\`,
}
`;

snapshot[`two headings 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"></div>
    </section>
  
    <section>
      <section-heading
        id='id-2'
        level=2
      >
        2
      </section-heading>
      <div class="section-body"><p>a</p>
</div>
    </section>
  
    <section>
      <section-heading
        id='id-3'
        level=2
      >
        3
      </section-heading>
      <div class="section-body"><p>b</p>
</div>
    </section>
  </div>\`,
}
`;

snapshot[`h2-6 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"></div>
    </section>
  
    <section>
      <section-heading
        id='id-2'
        level=2
      >
        2
      </section-heading>
      <div class="section-body"><p>a</p>
</div>
    </section>
  
    <section>
      <section-heading
        id='id-3'
        level=3
      >
        3
      </section-heading>
      <div class="section-body"><p>b
c</p>
</div>
    </section>
  
    <section>
      <section-heading
        id='id-4'
        level=4
      >
        4
      </section-heading>
      <div class="section-body"><p>d
e
f</p>
</div>
    </section>
  
    <section>
      <section-heading
        id='id-5'
        level=5
      >
        5
      </section-heading>
      <div class="section-body"><p>g
h
i
j</p>
</div>
    </section>
  
    <section>
      <section-heading
        id='id-6'
        level=6
      >
        6
      </section-heading>
      <div class="section-body"><p>k
l
m
n
o</p>
</div>
    </section>
  </div>\`,
}
`;

snapshot[`fence 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"><pre>a
b
c</pre></div>
    </section>
  </div>\`,
}
`;

snapshot[`indented fence 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"><pre>a
b
c</pre></div>
    </section>
  </div>\`,
}
`;

snapshot[`doubly indented fence 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"><pre>  a
  b
  c</pre></div>
    </section>
  </div>\`,
}
`;

snapshot[`fence with hashes 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"><pre>## a
## b
## c</pre></div>
    </section>
  </div>\`,
}
`;

snapshot[`heading before fence 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"></div>
    </section>
  
    <section>
      <section-heading
        id='id-2'
        level=2
      >
        2
      </section-heading>
      <div class="section-body"><p>a</p>
<pre>b
c
d</pre></div>
    </section>
  </div>\`,
}
`;

snapshot[`heading after fence 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"><pre>a
b
c</pre></div>
    </section>
  
    <section>
      <section-heading
        id='id-2'
        level=2
      >
        2
      </section-heading>
      <div class="section-body"><p>d</p>
</div>
    </section>
  </div>\`,
}
`;

snapshot[`heading, fence, heading 1`] = `
Promise {
  \`<div class='markdown'>
    <section>
      
      <page-title
        modified='2000-01-02'
        label='Title'
        published='2000-01-01'
        sublabel='Headline.'></page-title>
    
      <div class="section-body"></div>
    </section>
  
    <section>
      <section-heading
        id='id-2'
        level=2
      >
        2
      </section-heading>
      <div class="section-body"><p>a</p>
<pre>b
c
d</pre></div>
    </section>
  
    <section>
      <section-heading
        id='id-3'
        level=3
      >
        3
      </section-heading>
      <div class="section-body"><p>e</p>
</div>
    </section>
  </div>\`,
}
`;
